### PR TITLE
feat(ls-search): Search with patterns in Language Server

### PR DESCRIPTION
**Pull request by ajbt200128**
Added search functionality, and also made it so the server actually exits when requested to.

Sister PR [here](https://github.com/returntocorp/semgrep-vscode/pull/42)

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: src/language_server/custom_requests/Search.ml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    We then try and parse the pattern in every language (or specified lang), and
+    scan like normal, only returning the match ranges per file *)
+let on_request config targets params =
+  match params with
+  | None -> None
+  | Some params ->
+      let params = Structured.yojson_of_t params in
+      let pattern = params |> member "pattern" |> to_string in
+      let lang_opt = params |> member "language" |> to_string_option in
+      let lang_opt = Option.bind lang_opt Lang.of_string_opt in
```
Comment (brandonspark) [inline src/language_server/custom_requests/Search.ml:153]: let*, perhaps?

File: src/language_server/custom_requests/Search.ml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    let rule = Rule.rule_of_xpattern xlang xpat in
+    { rule with id = (Rule.ID.of_string "-", fk) }
+  in
+  match lang_opt with
+  | Some lang ->
+      (* TODO? capture also parse errors here? and transform the pattern
+       * parse error in invalid_rule_error to return in rules_and_origin? *)
+      [ rule_for_xlang (Xlang.of_lang lang) ]
+  | None ->
+      (* We need uniq_by because Lang.assoc contain multiple times the
```
Comment (brandonspark) [inline src/language_server/custom_requests/Search.ml:62]: i see you've copied this from `Rule_fetching.ml` -- any chance that we could extract this logic to a single helper, or if that's too far beyond the scope of this PR, leave a comment coupling the two?

**Review by brandonspark — APPROVED** [review]
just some small stuff

File: src/language_server/custom_requests/Search.ml (hunk 2 above)
Comment (ajbt200128) [reply src/language_server/custom_requests/Search.ml:62]: There's a comment on line 40 re: this :)

File: src/language_server/custom_requests/Search.ml (hunk 1 above)
Comment (ajbt200128) [reply src/language_server/custom_requests/Search.ml:153]: Did not know this existed, thank you!

File: src/language_server/custom_requests/Search.ml (hunk 2 above)
Comment (brandonspark) [reply src/language_server/custom_requests/Search.ml:62]: so you did! completely missed that

**Pull request merged**